composition/src/index.js

```diff
@@ -0,0 +1,258 @@
+import { options } from 'preact';
+
+/** @type {import('./internal').Component} */
+let currentComponent;
+
+let oldBeforeRender = options._render;
+options._render = vnode => {
+	if (oldBeforeRender) oldBeforeRender(vnode);
+
+	const c = (currentComponent = vnode._component);
+
+	const isFirst = !c.__compositions;
+	if (isFirst && c.constructor.__compositions) {
+		c.__compositions = { u: [], w: [], e: [], x: {} };
+		c.constructor = c.constructor(getterProxy((_, p) => c.props[p]));
+	}
+
+	if (c.__compositions && !isFirst)
+		// call all onMounted lifecycle callbacks
+		c.__compositions.w.forEach(up => handleEffect(up, c));
+};
+
+let oldAfterDiff = options.diffed;
+options.diffed = vnode => {
+	if (oldAfterDiff) oldAfterDiff(vnode);
+
+	const c = vnode._component;
+	if (c && c.__compositions)
+		// handle all `effect`s
+		c.__compositions.e.forEach(up => handleEffect(up, c));
+};
+
+let oldBeforeUnmount = options.unmount;
+options.unmount = vnode => {
+	if (oldBeforeUnmount) oldBeforeUnmount(vnode);
+
+	const c = vnode._component;
+	if (c && c.__compositions) {
+		// cleanup `effect`s onCleanup
+		c.__compositions.e.forEach(cleanupEffect);
+		// call all onUnmounted lifecycle callbacks
+		c.__compositions.u.forEach(f => f());
+	}
+};
+
+const $Reactive = Symbol('reactive');
+
+export function createComponent(comp) {
+	comp.__compositions = true;
+	return comp;
+}
+
+export function watch(src, cb, dv) {
+	const vr = { value: dv };
+	Object.defineProperty(vr, $Reactive, {
+		get() {
+			return this.value;
+		}
+	});
+	const up = { src, cb, vr };
+	handleEffect(up, currentComponent);
+	currentComponent.__compositions.w.push(up);
+	return vr;
+}
+
+export function effect(src, cb) {
+	currentComponent.__compositions.e.push({ src, cb });
+}
+
+export function onMounted(cb) {
+	currentComponent.__compositions.e.push({ cb });
+}
+
+export function onUnmounted(cb) {
+	currentComponent.__compositions.u.push(cb);
+}
+
+export function provide(name, _value) {
+	const _component = currentComponent;
+	if (!_component.__compositions.c) {
+		_component.__compositions.c = {};
+		_component.getChildContext = () => _component.__compositions.c;
+	}
+	_component.__compositions.c[`__sC_${name}`] = { _component, _value };
+}
+
+export function inject(name, defaultValue) {
+	const c = currentComponent;
+
+	const ctx = c.context[`__sC_${name}`];
+	if (!ctx) return defaultValue;
+
+	const src = ctx._value;
+
+	if (isReactive(src))
+		ctx._component.__compositions.w.push({ src, cb: () => c.setState({}) });
+
+	return src;
+}
+
+export function reactive(value) {
+	const c = currentComponent;
+
+	return new Proxy(
```

Comment:
I don't think proxies support our current  browser support

Reply:
well, could add it with a note that `preact/composition` only works with Proxies

Reply:
I will try to remove the need for Proxies, so no need for polyfill, and will not change browser support, hopefully

See the comment below for more information

Reply:
Removed Proxy from the PR